Sources/SIL/SExpr.swift

```diff
@@ -0,0 +1,126 @@
+// This is a temporary addition to libSIL and will likely be removed in the future.
+// It is a hack that allows one to read the output of e.g. swiftc -dump-ast.
+public enum SExpr : Equatable {
```

Comment:
How about a rebranding given that these data structures aren't s-expressions in the traditional sense of the word: https://en.wikipedia.org/wiki/S-expression?

Here's a tentative design that focuses on the fact that we're parsing ASTs as opposed to a general-purpose s-expression like data structure:

```
struct DumpedTree {
  let kind: String
  let fields: [DumpedField]
}

struct DumpedField {
  let name: String?
  let value: DumpedValue
}

enum DumpedValue {
  case string(String)
  case sourceRange(String)
  case tree(DumpedTree)
}
```

Reply:
I don't see how they are not s-expressions in the traditional sense of this word?

Also, while I don't mind changing the model, I'd like to see more concrete arguments than "I'd do it that way". I don't really see how the proposed change is more AST-specific except for the explicit use of the word `Tree` in one type. I'd rather keep what's there (which I'm already using in my implementation) unless you see any significant flaws in the current design.

Reply:
We had a chat about that in-person, and I'll summarize the discussion.

I'd like to see the data model to be more worked out. E.g. concretely:
  * To have `parse` return `Tree`, not `SExpr`.
  * To avoid having `.record` in favor of more precise data types like `Tree` and whatever else gets dumped as nested s-expr like things.

Also, I don't think that something that has records and properties qualifies as an s-expression. It's convenient to call this data format "s-expression" and forget about this inconsistency, but that's another aspect of working out the data model.

Given that this is a research project and that you're currently the only user of this API, I'm reluctantly okay with not making this a blocker for a merge.